src/compose/service.ts

```diff
@@ -702,24 +725,70 @@ export class Service {
 				log.debug('  Network changes detected');
 			}
 		}
-		return sameNetworks && sameConfig;
+
+		// Check the network mode separetely, as if it is a
+		// service: network mode, the container id needs to be
+		// checked against the running containers
+
+		// When this function is called, it's with the current
+		// state as a parameter and the target as the instance.
+		// We shouldn't rely on that because it's not enforced
+		// anywhere. For that reason we need to consider both
+		// network_modes in the correct way
+		let sameNetworkMode = true;
+		for (const [a, b] of [
+			[this.config.networkMode, service.config.networkMode],
+			[service.config.networkMode, this.config.networkMode],
+		]) {
+			const aMatch = a.match(SERVICE_NETWORK_MODE_REGEX);
+			const bMatch = b.match(SERVICE_NETWORK_MODE_REGEX);
+			if (aMatch != null) {
+				if (bMatch != null) {
+					// They're both service: <name> network modes
+					if (aMatch[1] !== bMatch[1]) {
+						sameNetworkMode = sameNetworkMode || false;
```

Comment:
What I was going for here is that once `sameNetworkMode` goes to `true`, it never changes again (because we check both 'orders' of the network mode. I can see how this code could be confusing though, I'll look into making it a little more obvious.

Reply:
I would suggest adding a unit test that exercises different variants and includes both equal and not-equal cases (right now, you only have a test for `isEqualConfig.to.be.true`). I still don't feel this implementation does what it should.